dali/test/dali_cuda_finalize_test.h

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_TEST_DALI_CUDA_FINALIZE_TEST_H_
+#define DALI_TEST_DALI_CUDA_FINALIZE_TEST_H_
+
+#include <cuda_runtime.h>
+#include <gtest/gtest.h>
+#include "dali/core/cuda_utils.h"
+
+namespace dali {
+
+/**
+ * @brief GoogleTest Event Listener, that checks after every Test Case if there were
+ * no CUDA related errors.
+ *
+ * We don't want to use TearDown methods that would need to be either injected into every
+ * Fixture or defined globally and would run once. Events allow us to define additional
+ * checks with any granularity we want.
+ *
+ * We can use GTest assertion macros in the events with exception of OnTestPartResult.
+ */
+class CudaFinalizeEventListener : public ::testing::EmptyTestEventListener {
+  void OnTestEnd(const ::testing::TestInfo& test_info) override {
+    auto sync_result = cudaDeviceSynchronize();
```

Comment:
`sync_result` is always `cudaSuccess` IIRC

Also I'm wondering, could this actually catch the error from another thread or this synchronization secures this scenario

Reply:
The docs is pretty clear:
```
Blocks until the device has completed all preceding requested tasks. cudaDeviceSynchronize() returns an error if one of the preceding tasks has failed.
```

Reply:
It theoretically can catch error from other thread (apparently use of threading is on by default), but this sync should limit the leaking of error to something close. And I observed a scenario when cudaDeviceSynchronize returned success when there was launch error and a case where both calls failed.

Reply:
I think the error is reported only once. If the launch error appeared then it won't be reported again. @mzient

Reply:
I forced a kernel launch failure and cudaDeviceSynchronize didn't report it, but the cudaGetLastError did (and this clears it for further calls).